lib/dhcp_wire.ml

```diff
@@ -953,73 +953,52 @@ let buf_of_options sbuf options =
     set_uint8 ebuf 0 (option_code_to_int END); shift ebuf 1
 
 let pkt_of_buf buf len =
-  let wrap () =
-    let open Wire_structs in
-    let open Ipv4_wire in
-    let open Printf in
-    let min_len = sizeof_dhcp + sizeof_ethernet + sizeof_ipv4 + sizeof_udp in
-    if len < min_len then
-      invalid_arg (sprintf "packet is too small: %d < %d" len min_len);
-    (* Handle ethernet *)
-    let srcmac = Macaddr.of_bytes_exn (copy_ethernet_src buf) in
-    let dstmac = Macaddr.of_bytes_exn (copy_ethernet_dst buf) in
-    let () = if (get_ethernet_ethertype buf) <> 0x0800 then
-        invalid_arg (sprintf "packet is not ipv4: %d" (get_ethernet_ethertype buf));
-    in
-    let buf = Cstruct.shift buf sizeof_ethernet in
-    (* Handle IPv4 *)
-    let ihl = (get_ipv4_hlen_version buf land 0xf) * 4 in
-    let ipcsum = get_ipv4_csum buf in
-    let csum = Tcpip_checksum.ones_complement (Cstruct.sub buf 0 ihl) in
-    (* Some broken clients don't do ip checksum, accept if they send as zero *)
-    let () = if ipcsum <> 0 && csum <> 0 then
-        invalid_arg (sprintf "bad ip checksum: 0x%x" ipcsum)
-    in
-    let () = if (get_ipv4_proto buf) <> 17 then
-        invalid_arg (sprintf "packet is not udp: %d" (get_ipv4_proto buf));
-    in
-    let srcip = Ipaddr.V4.of_int32 (get_ipv4_src buf) in
-    let dstip = Ipaddr.V4.of_int32 (get_ipv4_dst buf) in
-    let buf = Cstruct.shift buf ihl in
-    (* Handle UDP *)
-    let srcport = get_udp_source_port buf in
-    let dstport = get_udp_dest_port buf in
-    let buf = Cstruct.shift buf sizeof_udp in
-    (* Get the DHCP stuff *)
-    let op = int_to_op_exn (get_dhcp_op buf) in
-    let htype = if (get_dhcp_htype buf) = 1 then Ethernet_10mb else Other in
-    let hlen = get_dhcp_hlen buf in
-    let hops = get_dhcp_hops buf in
-    let xid = get_dhcp_xid buf in
-    let secs = get_dhcp_secs buf in
-    let flags =
-      if ((get_dhcp_flags buf) land 0x8000) <> 0 then Broadcast else Unicast
-    in
-    let ciaddr = Ipaddr.V4.of_int32 (get_dhcp_ciaddr buf) in
-    let yiaddr = Ipaddr.V4.of_int32 (get_dhcp_yiaddr buf) in
-    let siaddr = Ipaddr.V4.of_int32 (get_dhcp_siaddr buf) in
-    let giaddr = Ipaddr.V4.of_int32 (get_dhcp_giaddr buf) in
-    let chaddr =
+  let open Rresult in
+  let open Printf in
+  let min_len = sizeof_dhcp + Ethif_wire.sizeof_ethernet + Ipv4_wire.sizeof_ipv4 + Udp_wire.sizeof_udp in
+  if len < min_len then
+    invalid_arg (sprintf "packet is too small: %d < %d" len min_len);
+  (* Handle ethernet *)
+  Ethif_packet.Unmarshal.of_cstruct buf >>= fun (eth_header, eth_payload) ->
```

Comment:
Yes, that's `Rresult.bind` - similar monadic goodness to `Lwt.bind`, but for `Result.result`.